huaweicloud/resource_huaweicloud_cbr_vault.go

```diff
@@ -0,0 +1,431 @@
+		opts.Name = d.Get("name").(string)
+		opts.Billing = &vaults.BillingUpdate{
+			Size: d.Get("size").(int),
+		}
+		ae := d.Get("auto_expand").(bool)
+		opts.AutoExpand = &ae
+	}
+
+	if opts != (vaults.UpdateOpts{}) {
+		_, err := vaults.Update(client, d.Id(), opts).Extract()
+		if err != nil {
+			return fmt.Errorf("Error updating the vault: %s", err)
+		}
+	}
+
+	if d.HasChange("resources") {
+		if err := updateResources(d, client); err != nil {
+			return err
+		}
+	}
+	if d.HasChange("policy_id") {
+		if err := updatePolicy(d, client); err != nil {
+			return err
+		}
+	}
+
+	return resourceCBRVaultV3Read(d, meta)
+}
+
+func resourceCBRVaultV3Delete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	client, err := config.CbrV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating Huaweicloud CBR v3 client: %s", err)
+	}
+
+	if err := vaults.Delete(client, d.Id()).ExtractErr(); err != nil {
+		return fmt.Errorf("Error deleting CBR v3 vault: %s", err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourceIncludeVolume(volumes []interface{}) []vaults.ResourceExtraInfoIncludeVolumes {
+	includeVolumes := make([]vaults.ResourceExtraInfoIncludeVolumes, len(volumes))
+	if len(volumes) == 0 {
+		return includeVolumes
+	}
+	for i, v := range volumes {
+		includeVolumes[i] = vaults.ResourceExtraInfoIncludeVolumes{
+			ID: v.(string),
+		}
+	}
+	return includeVolumes
+}
+
+func resourceExcludeVolume(volumes []interface{}) []string {
+	includeVolumes := make([]string, len(volumes))
+	if len(volumes) == 0 {
+		return includeVolumes
+	}
+	for i, v := range volumes {
+		includeVolumes[i] = v.(string)
+	}
+	return includeVolumes
+}
+
+func buildCBRVaultResources(d *schema.ResourceData) []vaults.ResourceCreate {
+	res := make([]vaults.ResourceCreate, 0)
+	resType := d.Get("type").(string)
+	for _, v := range d.Get("resources").(*schema.Set).List() {
+		resourceID := v.(map[string]interface{})["id"].(string)
+		resourceType := resourceType[resType]
+		includes := resourceIncludeVolume(v.(map[string]interface{})["include_volumes"].([]interface{}))
+		excludes := resourceExcludeVolume(v.(map[string]interface{})["exclude_volumes"].([]interface{}))
+		extra := vaults.ResourceExtraInfo{
+			IncludeVolumes: includes,
+			ExcludeVolumes: excludes,
+		}
+		res = append(res, vaults.ResourceCreate{
+			ID:        resourceID,
+			Type:      resourceType,
+			ExtraInfo: &extra,
+		})
+	}
+	return res
+}
+
+func buildCBRVaultBilling(d *schema.ResourceData) *vaults.BillingCreate {
+	billing := &vaults.BillingCreate{
+		ObjectType:      d.Get("type").(string),
+		ConsistentLevel: d.Get("consistent_level").(string),
+		ProtectType:     d.Get("protection_type").(string),
+		Size:            d.Get("size").(int),
+	}
+
+	return billing
+}
+
+//Take out and save all the IDs nested in the list of resources map as a new string array.
+func changeSliceElementType(slice []interface{}) []string {
```

Comment:
please define the function with a better name

Reply:
The function name has been updated to `getResourceIDs`.